BUG: Inconsistent behavior for `step` with slice for label-based indexing

### Pandas version checks

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.


### Reproducible Example

```python
# /// script
# requires-python = ">=3.13"
# dependencies = [
#     "pandas==2.3.3",
# ]
# ///

import numpy as np
import pandas as pd

# Create a Series with non-contiguous integer index (step of 5)
# Index: 0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, ...
# Values: 0, 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, ...
T = np.arange(0, 100, 5)
series = pd.Series(np.arange(len(T)), index=T)

# Without step: returns all labels from 10 to 50 inclusive (label-based)
result_no_step = series.loc[10:50]
print("series.loc[10:50] (no step):")
print(series.loc[10:50])
print()

# With step=1: same as no step
print("series.loc[10:50:1] (step=1):")
print(series.loc[10:50:1])
print()

# With step=2: step is applied positionally, start, stop applied to labels
print("series.loc[10:50:2] (step=2):")
print(series.loc[10:50:2])
print()

# With step=5: Same behavior as step 2
print("series.loc[10:50:5] (step=5):")
print(series.loc[10:50:5])
print()

# Using arange with same arguments as the slice
print("series.loc[np.arange(10,50,5)] (step=5):")
print(series.loc[np.arange(10, 50, 5)])
print()
```

### Issue Description

When using `.loc` with a slice   `start/stop` are applied over a different space than `step` which I found very counterintuitive. I also was not able to find any docs (on this admittedly niche use case)

`start`/`stop` are applied over the values of the labels
`step` is applied positionally over the index

The result of the above script is

```
series.loc[10:50] (no step):
10     2
15     3
20     4
25     5
30     6
35     7
40     8
45     9
50    10
dtype: int64

series.loc[10:50:1] (step=1):
10     2
15     3
20     4
25     5
30     6
35     7
40     8
45     9
50    10
dtype: int64

series.loc[10:50:2] (step=2):
10     2
20     4
30     6
40     8
50    10
dtype: int64

series.loc[10:50:5] (step=5):
10    2
35    7
dtype: int64

series.loc[np.arange(10,50,5)] (step=5):
10    2
15    3
20    4
25    5
30    6
35    7
40    8
45    9
dtype: int64
```

### Expected Behavior

I would have expected either of the following:

### error

Throw an error saying that step is ambiguous and cannot be used here. This seems to be the approach of IntervalIndex:
https://github.com/pandas-dev/pandas/blob/499c5d4dd52a8645bf96c39bad60613097e84c06/pandas/core/indexes/interval.py#L978-L982

(Though as a sidenote I wasn't able to hit that code path)

### Step applies to  Label Space

In my example I would expect the slice with `step=5` to behave the same as `step=1` as it should hit each of the same values. My mental model is that for the case of integers as in my example

`series.loc[slice(start, stop, step)]`

should be equivalent to

`series.loc[np.arange(start, stop+step, step)]`
(with a +step to account for pandas inclusivity of bounds in slicing)
and in more amgious cases e.g. `slice("a", "f", 3)` and error should be thrown


### Installed Versions

<details><summary>INSTALLED VERSIONS</summary>

```

------------------
commit                : 9c8bc3e55188c8aff37207a74f1dd144980b8874
python                : 3.13.0
python-bits           : 64
OS                    : Darwin
OS-release            : 24.6.0
Version               : Darwin Kernel Version 24.6.0: Mon Jul 14 11:30:51 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T8112
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.3.3
numpy                 : 2.3.5
pytz                  : 2025.2
dateutil              : 2.9.0.post0
pip                   : None
Cython                : None
sphinx                : None
IPython               : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
blosc                 : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
html5lib              : None
hypothesis            : None
gcsfs                 : None
jinja2                : None
lxml.etree            : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
psycopg2              : None
pymysql               : None
pyarrow               : None
pyreadstat            : None
pytest                : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
xlsxwriter            : None
zstandard             : None
tzdata                : 2025.2
qtpy                  : None
pyqt5                 : None
```
</details>

here is a more minimal reproducer:

```python
# /// script
# requires-python = ">=3.13"
# dependencies = [
#     "pandas==2.3.3",
# ]
# ///

import numpy as np
import pandas as pd

# Create a Series with non-contiguous integer index (step of 5)
# Index: 0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, ...
# Values: 0, 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, ...
T = np.arange(0, 100, 5)
series = pd.Series(np.arange(len(T)), index=T)

aranged = series.loc[np.arange(10, 50, 5)]
sliced = series.loc[10:50:5]

pd.testing.assert_series_equal(sliced, aranged)
```


gives:


```
➜   uv run pandas_simple.py
Traceback (most recent call last):
  File "/Users/ian/Documents/dev/xarray-issue-debug/pandas_simple.py", line 22, in <module>
    pd.testing.assert_series_equal(sliced, aranged)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/ian/.cache/uv/environments-v2/pandas-simple-7d0e8f0fa412c403/lib/python3.13/site-packages/pandas/_testing/asserters.py", line 960, in assert_series_equal
    raise_assert_detail(obj, "Series length are different", msg1, msg2)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ian/.cache/uv/environments-v2/pandas-simple-7d0e8f0fa412c403/lib/python3.13/site-packages/pandas/_testing/asserters.py", line 620, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: Series are different

Series length are different
[left]:  2, Index([10, 35], dtype='int64')
[right]: 8, Index([10, 15, 20, 25, 30, 35, 40, 45], dtype='int64')
```

--- a/doc/source/user_guide/indexing.rst
+++ b/doc/source/user_guide/indexing.rst
@@ -431,22 +431,34 @@ an error will be raised. For instance, in the above example, ``s.loc[2:5]`` woul
 For more information about duplicate labels, see
 :ref:`Duplicate Labels <duplicates>`.
 
-Also, when using a slice with a step, such as ``.loc[start:stop:step]``, note that
+When using a slice with a step, such as ``.loc[start:stop:step]``, note that
 *start* and *stop* are interpreted as **labels**, while *step* is applied over
 the **positional index** within that label range. This means a stepped slice
-may return different labels than selecting an explicit list, even when they
-appear similar.
+will behave differently than using the labels ``range(start, stop, step)`` when
+the index is not contiguous integers.
 
 For example, in a ``Series`` with a non-contiguous integer index:
 
 .. ipython:: python
 
-   s = pd.Series(range(10), index=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45])
-   s.loc[10:50:5]              # (10), then skip 3 positions → 35 only
-   s.loc[[10, 15, 20, 25]]     # explicit label selection
+    s = pd.Series(range(10), index=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45])
+    s.loc[10:50:5]              # (10), then skip 3 positions → 35 only
+    s.loc[[10, 15, 20, 25]]     # explicit label selection
 
 The first applies *step* across **positional locations** between the start/stop
 labels. The second selects each label directly.
+
+Similarly, with a string-based index, the behavior is identical:
+
+.. ipython:: python
+
+    s = pd.Series(range(10), index=['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'])
+    s.loc['b':'i':2]            # Start at 'b' (position 1), stop at 'i' (position 8), step 2 positions → 'b', 'd', 'f', 'h'
+    s.loc[['b', 'd', 'f', 'h']] # explicit label selection
+
+In both cases, *start* and *stop* determine the label boundaries (inclusive),
+while *step* skips positions within that range, regardless of the index type.
+
 
 
 .. _indexing.integer:

--- a/doc/source/user_guide/indexing.rst
+++ b/doc/source/user_guide/indexing.rst
@@ -431,6 +431,24 @@ an error will be raised. For instance, in the above example, ``s.loc[2:5]`` woul
 For more information about duplicate labels, see
 :ref:`Duplicate Labels <duplicates>`.
 
+Also, when using a slice with a step, such as ``.loc[start:stop:step]``, note that
+*start* and *stop* are interpreted as **labels**, while *step* is applied over
+the **positional index** within that label range. This means a stepped slice
+may return different labels than selecting an explicit list, even when they
+appear similar.
+
+For example, in a ``Series`` with a non-contiguous integer index:
+
+.. ipython:: python
+
+   s = pd.Series(range(10), index=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45])
+   s.loc[10:50:5]              # (10), then skip 3 positions → 35 only
+   s.loc[[10, 15, 20, 25]]     # explicit label selection
+
+The first applies *step* across **positional locations** between the start/stop
+labels. The second selects each label directly.
+
+
 .. _indexing.integer:
 
 Selection by position

